Zend/zend_builtin_functions.c

```diff
@@ -1696,61 +1683,67 @@ ZEND_FUNCTION(debug_print_backtrace)
 		RETURN_THROWS();
 	}
 
-	ZVAL_UNDEF(&arg_array);
-	ptr = EX(prev_execute_data);
-
 	/* skip debug_backtrace() */
-	call = ptr;
-	ptr = ptr->prev_execute_data;
+	call = EX(prev_execute_data);
+
+	while (call && (limit == 0 || frameno < limit)) {
+		zend_execute_data *prev = call->prev_execute_data;
+
+		if (!prev) {
+			/* add frame for a handler call without {main} code */
+			if (EXPECTED((ZEND_CALL_INFO(call) & ZEND_CALL_TOP_FUNCTION) == 0)) {
+				break;
+			}
+		} else if (UNEXPECTED((ZEND_CALL_INFO(call) & ZEND_CALL_GENERATOR) != 0)) {
+			prev = zend_generator_check_placeholder_frame(prev);
+		}
 
-	while (ptr && (limit == 0 || frameno < limit)) {
 		frameno++;
 		class_name = NULL;
 		call_type = NULL;
 		ZVAL_UNDEF(&arg_array);
 
-		if (UNEXPECTED((ZEND_CALL_INFO(call) & ZEND_CALL_GENERATOR) != 0)) {
-			ptr = zend_generator_check_placeholder_frame(ptr);
-		}
-
-		skip = ptr;
-		/* skip internal handler */
-		if (skip_internal_handler(skip)) {
-			skip = skip->prev_execute_data;
-		}
-
-		if (skip->func && ZEND_USER_CODE(skip->func->common.type)) {
-			filename = ZSTR_VAL(skip->func->op_array.filename);
-			if (skip->opline->opcode == ZEND_HANDLE_EXCEPTION) {
+		if (prev && prev->func && ZEND_USER_CODE(prev->func->common.type)) {
+			filename = ZSTR_VAL(prev->func->op_array.filename);
+			if (prev->opline->opcode == ZEND_HANDLE_EXCEPTION) {
 				if (EG(opline_before_exception)) {
 					lineno = EG(opline_before_exception)->lineno;
 				} else {
-					lineno = skip->func->op_array.line_end;
+					lineno = prev->func->op_array.line_end;
 				}
 			} else {
-				lineno = skip->opline->lineno;
+				lineno = prev->opline->lineno;
 			}
 		} else {
+			zend_execute_data *prev_call = prev;
+
+			while (prev_call) {
+				zend_execute_data *prev;
+
+				if (prev_call &&
+					prev_call->func &&
+					!ZEND_USER_CODE(prev_call->func->common.type) &&
+					!(prev_call->func->common.fn_flags & ZEND_ACC_CALL_VIA_TRAMPOLINE)) {
+					break;
+				}
+
+				prev = prev_call->prev_execute_data;
+				if (prev && prev->func && ZEND_USER_CODE(prev->func->common.type)) {
+					filename = ZSTR_VAL(prev->func->op_array.filename);
+					lineno = prev->opline->lineno;
```

Comment:
This code sets filename and lineno, breaks, and then below filename and lineno will be unconditionally overwritten with zero.

It seems like we have really bad test coverage for this code. I commented out the while while(prev_call) loop and nothing failed :(

Reply:
Good catch. This loop was taken from zend_fetch_debug_backtrace() (for consistency), but it doesn't make any effect because  ``filename`` and ``lineno`` are rest below. This reset should be done before the loop.